doc/sr_sender.1.rst
==========
 SR_Sender 
==========

-----------------------------------
Send posted files to remote servers
-----------------------------------

:Manual section: 1 
:Date: @Date@
:Version: @Version@
:Manual group: MetPX-Sarracenia

.. contents::

SYNOPSIS
========

**sr_sender** foreground|start|stop|restart|reload|status configfile

**sr_sender** cleanup|declare|setup configfile

DESCRIPTION
===========

**sr_sender** is a component derived from `sr_subscribe(1) <sr_subscribe.1.rst>`_
used to send local files to a remote server using a file transfer protocol, primarily SFTP.
**sr_sender** is a standard consumer, using all the normal AMQP settings for brokers, exchanges,
queues, and all the standard client side filtering with accept, reject, and on_message.

Often, a broker will announce files using a remote protocol such as HTTP,
but for the sender it is actually a local file.  In such cases, one will
see a message: **ERROR: The file to send is not local.** 
An on_message plugin will convert the web url into a local file one::

  base_dir /var/httpd/www
  on_message msg_2localfile

This on_message plugin is part of the default settings for senders, but one
still needs to specify base_dir for it to function.

If a **post_broker** is set, **sr_sender** checks if the clustername given
by the **to** option if found in one of the message's destination clusters.
If not, the message is skipped.


DESTINATION UNAVAILABLE
-----------------------

If the server to which the files are being sent is going to be unavailable for 
a prolonged period, and there is a large number of messages to send to it, then
the queue will build up on the broker. As the performance of the entire broker
is affected by large queues, one needs to minimize such queues.

The *-save* and *-restore* options are used get the messages away from the broker
when a very large a queue will certainly build up.
The *-save* option copies the messages to a (per instance) disk file (in the same directory
that stores state and pid files), as json encoded strings, one per line.
When a queue is building up::

   sr_sender stop <config> 
   sr_sender -save start <config> 

And run the sender in *save* mode (which continually writes incoming messages to disk)
in the log, a line for each message written to disk::

  2017-03-03 12:14:51,386 [INFO] sr_sender saving 2 message 
       topic: v02.post.home.peter.sarra_devdocroot.sub.SASP34_LEMM_031630__LEDA_60215

Continue in this mode until the absent server is again available.  At that point::

   sr_sender stop <config> 
   sr_sender -restore start <config> 

While restoring from the disk file, messages like the following will appear in the log::

  2017-03-03 12:15:02,969 [INFO] sr_sender restoring message 29 of 34: 
    topic: v02.post.home.peter.sarra_devdocroot.sub.ON_02GD022_daily_hydrometric.csv


After the last one::

  2017-03-03 12:15:03,112 [INFO] sr_sender restore complete deleting save 
    file: /home/peter/.cache/sarra/sender/tsource2send/sr_sender_tsource2send_0000.save 


and the sr_sender will function normally thereafter.



SETUP 1 : PUMP TO PUMP REPLICATION 
----------------------------------

 - **mirror             <boolean>   (default: True)** 
 - **simulate             <boolean>   (default: False)** 
 - **base_dir      <directory> (None)** 

 - **destination        <url>       (MANDATORY)** 
 - **do_send            <script>    (None)** 
 - **kbytes_ps          <int>       (default: 0)** 
 - **post_base_dir <directory> (default: '')** 

 - **to               <clustername> (default: <post_broker host>)** 
 - **on_post           <script>     (default: None)** 
 - **post_broker        amqp{s}://<user>:<pw>@<brokerhost>[:port]/<vhost>**
 - **url                <url>       (default: destination)** 

For pump replication, **mirror** is set to True (default).

**base_dir** supplies the directory path that, when combined with the relative
one in the selected notification gives the absolute path of the file to be sent.
The default is None which means that the path in the notification is the absolute one.

The **destination** defines the protocol and server to be used to deliver the products.
Its form is a partial url, for example:  **ftp://myuser@myhost**
The program uses the file ~/.conf/sarra/credentials.conf to get the remaining details
(password and connection options).  Supported protocol are ftp, ftps and sftp. Should the
user need to implement another sending mechanism, he would provide the plugin script 
through option **do_send**.

On the remote site, the **post_base_dir** serves the same purpose as the
**base_dir** on this server.  The default is None which means that the delivered path
is the absolute one.

Now we are ready to send the product... for example, if the selected notification looks like this :

**20150813161959.854 http://this.pump.com/ relative/path/to/IMPORTANT_product**

**sr_sender**  performs the following pseudo-delivery:

Sends local file [**base_dir**]/relative/path/to/IMPORTANT_product
to    **destination**/[**post_base_dir**]/relative/path/to/IMPORTANT_product
(**kbytes_ps** is greater than 0, the process attempts to respect 
this delivery speed... ftp,ftps,or sftp)

At this point, a pump-to-pump setup needs to send the remote notification...
(If the post_broker is not set, there will be no posting... just products replication)

The selected notification contains all the right information
(topic and header attributes) except for url field in the
notice... in our example :  **http://this.pump.com/**

By default, **sr_sender** puts the **destination** in that field. 
The user can overwrite this by specifying the option **post_base_url**. For example:

**post_base_url http://remote.apache.com**

The user can provide an **on_post** script. Just before the message is
published on the **post_broker**  and **post_exchange**, the 
**on_post** script is called... with the **sr_sender** class instance as an argument.
The script can perform whatever you want... if it returns False, the message will not 
be published. If True, the program will continue processing from there.  

**simulate**  replaces the active sender with one that just logs the name of the file
it would be sent with.


DESTINATION SETUP 2 : METPX-SUNDEW LIKE DISSEMINATION
-----------------------------------------------------

In this type of usage, we would not usually repost... but if the 
**post_broker** and **post_exchange** (**url**,**on_post**) are set,
the product will be announced (with its possibly new location and new name).
Let's reintroduce the options in a different order 
with some new ones to ease explanation.


 - **mirror             <boolean>   (default: True)** 
 - **base_dir      <directory> (None)** 

 - **destination        <url>       (MANDATORY)** 
 - **post_base_dir <directory> (default: '')** 

 - **directory          <path>      (MANDATORY)** 
 - **on_message            <script> (default: None)** 
 - **accept        <regexp pattern> (default: None)** 
 - **reject        <regexp pattern> (default: None)** 

There are 2 differences with the previous case : 
the **directory**, and the **filename** options.

The **base_dir** is the same, and so are the
**destination**  and the **post_base_dir** options.

The **directory** option defines another "relative path" for the product
at its destination.  It is tagged to the **accept** options defined after it.
If another sequence of **directory**/**accept** follows in the configuration file,
the second directory is tagged to the following accepts and so on.

The  **accept/reject**  patterns apply to message notice url as above.
Here is an example, here some ordered configuration options :

::

  directory /my/new/important_location

  accept .*IMPORTANT.*

  directory /my/new/location/for_others

  accept .*

If the notification selected is, as above, this :

**20150813161959.854 http://this.pump.com/ relative/path/to/IMPORTANT_product**

It was selected by the first **accept** option. The remote relative path becomes
**/my/new/important_location** ... and **sr_sender**  performs the following pseudo-delivery:

sends local file [**base_dir**]/relative/path/to/IMPORTANT_product
to    **destination**/[**post_base_dir**]/my/new/important_location/IMPORTANT_product


Usually this way of using **sr_sender** would not require posting of the product.
But if **post_broker** and **post_exchange** are provided, and **url** , as above, is set to
**http://remote.apache.com**,  then **sr_sender** would reconstruct :

Topic: **v02.post.my.new.important_location.IMPORTANT_product**

Notice: **20150813161959.854 http://remote.apache.com/ my/new/important_location/IMPORTANT_product**


SEE ALSO
========

`sr_subscribe(1) <sr_subscribe.1.rst>`_ - the format of configurations for MetPX-Sarracenia.

`sr_report(7) <sr_report.7.rst>`_ - the format of report messages.

`sr_report(1) <sr_report.1.rst>`_ - process report messages.

`sr_post(1) <sr_post.1.rst>`_ - post announcemensts of specific files.

`sr_post(7) <sr_post.7.rst>`_ - the format of announcements.

`sr_subscribe(1) <sr_subscribe.1.rst>`_ - the download client.

`sr_watch(1) <sr_watch.1.rst>`_ - the directory watching daemon.

`dd_subscribe(1) <dd_subscribe.1.rst>`_ - the http-only download client.


CHANGES.rst
===============
 Release Notes 
===============

lists all changes between versions.

**git repo**

v03_wip ... still in development, too many changes to list.
unstable, breaking changes possible, only for experimental use.


sarra/sr_sender.py
#!/usr/bin/env python3
#
# This file is part of sarracenia.
# The sarracenia suite is Free and is proudly provided by the Government of Canada
# Copyright (C) Her Majesty The Queen in Right of Canada, Environment Canada, 2008-2015
#
# Questions or bugs report: dps-client@ec.gc.ca
# Sarracenia repository: https://github.com/MetPX/sarracenia
# Documentation: https://github.com/MetPX/sarracenia
#
# sr_sender.py : python3 program consumes product messages and send them to another pump
#                and announce the newly arrived product on that pump. If the post_broker
#                is not given... it is accepted, the products are sent without notices.
#
#
# Code contributed by:
#  Michel Grenier - Shared Services Canada
#  Last Changed   : Wed Oct  4 20:24 UTC 2017
#                   code rewritten : sr_sender is an instantiation of sr_subscribe
#
########################################################################
#  This program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; version 2 of the License.
#
#  This program is distributed in the hope that it will be useful, 
#  but WITHOUT ANY WARRANTY; without even the implied warranty of 
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the 
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program; if not, write to the Free Software
#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307  USA
#
#============================================================
# usage example
#
# sr_sender [options] [config] [foreground|start|stop|restart|reload|status|cleanup|setup]
#
# sr_sender connects to a broker. For each product it has selected, it sends it onto the other
# broker and reannounce it there.
#
# conditions:
#
# broker                  = where sarra is running (manager)
# exchange                = xpublic
# message.headers[to_clusters] verified if destination includes remote pump ?
#
# do_send                 = a script supporting the protocol defined in the destination
# destination             = an url of the credentials of the remote server and its options (see credentials)
# directory               = the placement is mirrored usually
# accept/reject           = pattern matching what we want to poll in that directory
#
# (optional... only if message are posted after products are sent)
# post_broker             = remote broker (remote manager)
# post_exchange           = xpublic
# document_root           = if provided, extracted from the url if present
# url                     = build from the destination/directory/product
# product                 = the product placement is mirrored by default
#                           unless if accept/reject are under a directory option
# post_message            = incoming message with url changes
#                           message.headers['source']  left as is
#                           message.headers['cluster'] left as is 
#                           option url : gives new url announcement for this product
#============================================================

#

import os,sys,time
import json

try :    
         from sr_subscribe       import *
except : 
         from sarra.sr_subscribe import *

class sr_sender(sr_subscribe):

    def check(self):
        self.sleep_connect_try_interval_min=0.01
        self.sleep_connect_try_interval_max=30
        self.sleep_connect_try_interval=self.sleep_connect_try_interval_min

        self.connected     = False
        if self.config_name == None : return

        self.check_consumer_options()

        # posting... discard not permitted
        if self.post_broker != None :

           if self.post_exchange == None and self.post_exchange_suffix :
              self.post_exchange = 'xs_%s' % self.post_broker.username + '_' + self.post_exchange_suffix

           # enforcing post_exchange

           if self.post_exchange == None :
              if self.exchange   != None :
                 self.post_exchange = self.exchange
                 self.logger.warning("use post_exchange to set exchange")

           # verify post_base_dir

           if self.post_base_dir == None :
              if self.post_document_root != None :
                 self.post_base_dir = self.post_document_root
                 self.logger.warning("use post_base_dir instead of post_document_root")
              elif self.document_root != None :
                 self.post_base_dir = self.document_root
                 self.logger.warning("use post_base_dir instead of defaulting to document_root")

           # verify post_base_url (not mandatory...)

           if self.post_base_url == None :
              self.post_base_url = self.destination
           # overwrite inflight to be None when posting to a broker.
           if self.inflight == 'unspecified':
              self.inflight=None
        else:
           if self.inflight == 'unspecified':
              self.inflight='.tmp'

        # check destination
        self.details = None
        if self.destination != None :
           ok, self.details = self.credentials.get(self.destination)

        #if self.destination == None or self.details == None:
        # destination is not required if a do_send is provided
        if self.do_send is None and (self.destination is None or self.details is None):
           self.logger.error("destination option incorrect or missing\n")
           sys.exit(1)

        # to clusters required

        if self.to_clusters == None and self.post_broker != None :
           self.to_clusters = self.post_broker.hostname

        # do_task should have doit_send for now... make it a plugin later
        # and the sending is the last thing that should be done

        if not self.doit_send in self.do_task_list :
           self.do_task_list.append(self.doit_send)

        # ===========================================================
        # some sr_subscribe options reset to match sr_sarra behavior
        # ===========================================================

        # always sends ...

        if self.notify_only :
           self.logger.error("sender notify_only True\n")
           sys.exit(1)

        # never discard

        if self.discard :
           self.logger.error("sender discard True\n")
           sys.exit(1)


    # =============
    # __do_send__
    # =============

    def __do_send__(self):
        self.logger.debug("sending/copying %s to %s " % ( self.msg.relpath, self.msg.new_dir ) )

        # try registered do_send first... might overload defaults

        # if the  destination is None, so will the details.url.scheme.
        # with scheme = None and destination = None, you go straight to do_send()
        scheme = None
        if self.destination is not None: scheme = self.details.url.scheme
        try:
                if   scheme in self.do_sends :
                     self.logger.debug("using registered do_send for %s" % scheme)
                     do_send = self.do_sends[scheme]
                     ok = do_send(self)
                     # if ok == None  it means that the scheme was one
                     # of the supported python one (sftp,ftp[s])
                     # and the plugin decided to go with the python defaults
                     if ok != None : return ok
        except:
                self.logger.debug('Exception details:', exc_info=True)

        # try supported hardcoded send
        try : 
                if   scheme in ['ftp','ftps']  :
                     if not hasattr(self,'ftp_link') :
                        self.ftp_link = ftp_transport()
                     ok = self.ftp_link.send(self)
                     return ok

                elif scheme == 'sftp' :
                     try    : from sr_sftp       import sftp_transport
                     except : from sarra.sr_sftp import sftp_transport
                     if not hasattr(self,'sftp_link') :
                        self.sftp_link = sftp_transport(self)
                     ok = self.sftp_link.send(self)
                     return ok


                # user defined send scripts
                # if many are configured, this one is the last one in config

                elif self.do_send :
                     ok = self.do_send(self)
                     print("do_send is being called")
                     print(ok)
                     return ok

        except :
                if self.reportback:
                    self.msg.report_publish(503, "Unable to process")
                self.logger.error("sender/__do_send__: could not send")
                self.logger.debug('Exception details: ', exc_info=True)

        # something went wrong

        if self.reportback:
           self.msg.report_publish(503,"Service unavailable %s" % self.msg.url.scheme)

        return False

    def overwrite_defaults(self):
        # a destination must be provided

        self.destination    = None
        self.currentDir     = None
        self.currentPattern = None

        # consumer defaults

        if hasattr(self,'manager'):
           self.broker = self.manager
        self.exchange  = 'xpublic'

        # most of the time we want to mirror product directory

        self.mirror      = True

        # add msg_2localfile to the on_message_list at the beginning

        self.execfile("on_message",'msg_2localfile')
        self.on_message_list.insert(0,self.on_message )

        # always sends ...

        self.notify_only = False

        # never discard

        self.discard = False

        # dont accept unmatch by default

        self.accept_unmatch = False



    # =============
    # doit_send  
    # =============

    def doit_send(self,parent=None):
        self.logger.debug("doit_send with %s '%s %s %s' %s" % \
            (self.msg.topic, self.msg.pubtime, self.msg.baseurl, \
             self.msg.relpath, self.msg.hdrstr))

        self.logger.debug("doit_send to: %s / %s " % ( self.msg.new_dir, self.msg.new_file)  )
        # the code of msg_2localfile could be put here...

        #=================================
        # impossible to send
        #=================================

        # checking the destination was provided first since it is now optional in this script
        if self.destination is not None and self.destination[:3] == 'ftp' :
            # 'i' cannot be supported by ftp/ftps
            # we cannot offset in the remote file to inject data
            #
            # FIXME instead of dropping the message
            # the inplace part could be delivered as 
            # a separate partfile and message set to 'p'
            if  self.msg.partflg == 'i':
                self.logger.error("ftp, inplace part file not supported")
                if self.reportback:
                   msg.report_publish(499,'ftp cannot deliver partitioned files')
                return False

        #=================================
        # check message for local file
        #=================================

        if self.msg.baseurl != 'file:' :
           self.logger.error("protocol should be 'file:' message ignored")
           return False

        #=================================
        # proceed to send :  has to work
        #=================================

        # N attempts to send

        i  = 1
        while i <= self.attempts:
              if i != 1:
                  self.logger.warning("sending again, attempt %d" % i)
              ok = self.__do_send__()
              if ok : break
              # dont force on retry 
              if self.msg.isRetry : break
              i = i + 1
               
        # if retry mode... do retry stuff
        if self.retry_mode :
           if ok : self.consumer.msg_worked()
           else  : self.consumer.msg_to_retry()

        # could not download ...

        if not ok: return False

        #=================================
        # publish our sending
        #=================================

        if self.post_broker :
           self.msg.set_topic(self.post_topic_prefix,self.msg.new_relpath)
           self.msg.set_notice(self.msg.new_baseurl,self.msg.new_relpath,self.msg.pubtime)
           self.__on_post__()
           if self.reportback:
               self.msg.report_publish(201,'Published')

        return True

# ===================================
# MAIN
# ===================================

def main():

    args,action,config,old = startup_args(sys.argv)
    sender = sr_sender(config,args,action)
    sender.exec_action(action,old)

    os._exit(0)

# =========================================
# direct invocation
# =========================================

if __name__=="__main__":
   main()
